flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/FileSystemCatalog.java

```diff
@@ -0,0 +1,298 @@
+import org.apache.flink.table.store.file.schema.SchemaManager;
+import org.apache.flink.table.store.file.schema.UpdateSchema;
+import org.apache.flink.util.function.RunnableWithException;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+
+import static org.apache.flink.table.catalog.GenericInMemoryCatalogFactoryOptions.DEFAULT_DATABASE;
+import static org.apache.flink.table.store.file.FileStoreOptions.PATH;
+
+/** A catalog implementation for {@link FileSystem}. */
+public class FileSystemCatalog extends TableStoreCatalog {
+
+    public static final String DB_SUFFIX = ".db";
+
+    public static final CatalogDatabaseImpl DUMMY_DATABASE =
+            new CatalogDatabaseImpl(Collections.emptyMap(), null);
+
+    private final FileSystem fs;
+    private final Path root;
+
+    public FileSystemCatalog(String name, Path root) {
+        this(name, root, DEFAULT_DATABASE.defaultValue());
+    }
+
+    public FileSystemCatalog(String name, Path root, String defaultDatabase) {
+        super(name, defaultDatabase);
+        this.root = root;
+        this.fs = uncheck(root::getFileSystem);
+        uncheck(() -> createDatabase(defaultDatabase, DUMMY_DATABASE, true));
+    }
+
+    @Override
+    public Optional<Factory> getFactory() {
+        return Optional.of(
+                FactoryUtil.discoverFactory(
+                        classLoader(), DynamicTableFactory.class, "table-store"));
+    }
+
+    @Override
+    public List<String> listDatabases() throws CatalogException {
+        List<String> databases = new ArrayList<>();
+        for (FileStatus status : uncheck(() -> fs.listStatus(root))) {
+            Path path = status.getPath();
+            if (status.isDir() && isDatabase(path)) {
+                databases.add(database(path));
+            }
+        }
+        return databases;
+    }
+
+    @Override
+    public CatalogDatabase getDatabase(String databaseName)
+            throws DatabaseNotExistException, CatalogException {
+        if (!databaseExists(databaseName)) {
+            throw new DatabaseNotExistException(getName(), databaseName);
+        }
+        return new CatalogDatabaseImpl(Collections.emptyMap(), "");
+    }
+
+    @Override
+    public boolean databaseExists(String databaseName) throws CatalogException {
+        return uncheck(() -> fs.exists(databasePath(databaseName)));
+    }
+
+    @Override
+    public void createDatabase(String name, CatalogDatabase database, boolean ignoreIfExists)
+            throws DatabaseAlreadyExistException, CatalogException {
+        if (database.getProperties().size() > 0) {
+            throw new UnsupportedOperationException(
+                    "Create database with properties is unsupported.");
+        }
+
+        if (database.getDescription().isPresent() && !database.getDescription().get().equals("")) {
```

Comment:
```suggestion
        if (database.getDescription().isPresent() && StringUtils.isNullOrWhitespaceOnly(database.getDescription().get())) {
```

Reply:
I think current version is OK, it should not be null.